### PR TITLE
QALora example

### PR DESCRIPTION
**Pull request by cyita**
## Description

Support QALora & example.
Based on https://github.com/intel-analytics/BigDL/pull/9378, https://github.com/intel-analytics/bigdl-qlora-alpaca/pull/4

### 1. Why the change?

Support QALora & example.

### 2. User API changes

Users can pass qa_lora=True in bigdl.llm.transformers.qlora.LoraConfig
```
config = LoraConfig(
        r=lora_r,
        lora_alpha=lora_alpha,
        target_modules=lora_target_modules,
        lora_dropout=lora_dropout,
        bias="none",
        task_type="CAUSAL_LM",
        qa_lora=True,
    )
```

### 4. How to test?
- [x] Manual test

### CODE REVIEW

**Comment by cyita** [comment]
Arc-02

defualt gradient_accumulation_steps = 16
Use 4 in test.

q4_1, loss=nan once optimizer step on arc

#### LoraConfig
```
LoraConfig(peft_type=<PeftType.LORA: 'LORA'>, auto_mapping=None, base_model_name_or_path=None, revision=None, task_type=
'CAUSAL_LM', inference_mode=False, r=64, target_modules=['k_proj', 'q_proj', 'o_proj', 'down_proj', 'v_proj', 'gate_proj
', 'up_proj'], lora_alpha=16, lora_dropout=0.0, fan_in_fan_out=False, bias='none', modules_to_save=None, init_lora_weigh
ts=True, layers_to_transform=None, layers_pattern=None, qa_lora=True)
```

#### TrainArgs
```
TrainingArguments(
_n_gpu=1,
adafactor=False,
adam8bit=False,
adam_beta1=0.9,
adam_beta2=0.999,
adam_epsilon=1e-08,
auto_find_batch_size=False,
bf16=False,
bf16_full_eval=False,
cache_dir=None,
data_seed=None,
dataloader_drop_last=False,
dataloader_num_workers=0,
dataloader_pin_memory=True,
ddp_backend=None,
ddp_broadcast_buffers=None,
ddp_bucket_cap_mb=None,
ddp_find_unused_parameters=None,
ddp_timeout=1800,
debug=[],
deepspeed=None,
disable_tqdm=False,
dispatch_batches=None,
do_eval=False,
do_mmlu_eval=False,
do_predict=False,
do_train=True,
eval_accumulation_steps=None,
eval_delay=0,
eval_steps=None,
evaluation_strategy=no,
fp16=False,
fp16_backend=auto,
fp16_full_eval=False,
fp16_opt_level=O1,
fsdp=[],
fsdp_config={'min_num_params': 0, 'xla': False, 'xla_fsdp_grad_ckpt': False},
fsdp_min_num_params=0,
fsdp_transformer_layer_cls_to_wrap=None,
full_determinism=False,
full_finetune=False,
generation_config=GenerationConfig {
  "max_new_tokens": 256
}
,
generation_max_length=None,
generation_num_beams=None,
gradient_accumulation_steps=4,
gradient_checkpointing=True,
gradient_checkpointing_kwargs=None,
greater_is_better=None,
group_by_length=True,
half_precision_backend=auto,
hub_always_push=False,
hub_model_id=None,
hub_private_repo=False,
hub_strategy=every_save,
hub_token=<HUB_TOKEN>,
ignore_data_skip=False,
include_inputs_for_metrics=False,
include_tokens_per_second=False,
jit_mode_eval=False,
label_names=None,
label_smoothing_factor=0.0,
learning_rate=0.0002,
length_column_name=length,
load_best_model_at_end=False,
local_rank=0,
log_level=passive,
log_level_replica=warning,
log_on_each_node=True,
logging_dir=./bigdl-qalora-alpaca/runs/Nov29_12-00-08_arda-arc02,
logging_first_step=False,
logging_nan_inf_filter=True,
logging_steps=10,
logging_strategy=steps,
lora_alpha=16,
lora_dropout=0.0,
lora_r=64,
lr_scheduler_type=constant,
max_grad_norm=0.3,
max_memory_MB=24000,
max_mmlu_samples=None,
max_steps=-1,
metric_for_best_model=None,
mmlu_dataset=mmlu-fs,
mmlu_source_max_len=2048,
mmlu_split=eval,
mp_parameters=,
neftune_noise_alpha=None,
no_cuda=False,
num_train_epochs=3,
optim=adamw_torch,
optim_args=None,
output_dir=./bigdl-qalora-alpaca,
overwrite_output_dir=False,
past_index=-1,
per_device_eval_batch_size=8,
per_device_train_batch_size=1,
predict_with_generate=False,
prediction_loss_only=False,
push_to_hub=False,
push_to_hub_model_id=None,
push_to_hub_organization=None,
push_to_hub_token=<PUSH_TO_HUB_TOKEN>,
ray_scope=last,
remove_unused_columns=False,
report_to=[],
resume_from_checkpoint=None,
run_name=./bigdl-qalora-alpaca,
save_on_each_node=False,
save_safetensors=True,
save_steps=250,
save_strategy=steps,
save_total_limit=40,
seed=42,
skip_memory_metrics=True,
sortish_sampler=False,
split_batches=False,
tf32=None,
torch_compile=False,
torch_compile_backend=None,
torch_compile_mode=None,
torchdynamo=None,
tpu_metrics_debug=False,
tpu_num_cores=None,
train_on_source=False,
use_cpu=False,
use_ipex=False,
use_legacy_prediction_loop=False,
use_mps_device=False,
warmup_ratio=0.03,
warmup_steps=0,
weight_decay=0.0,
)
```


#### Llama2-7b-chat-hf, q4_0, fp32
```
torch.float32 350580736 0.09247812327463165
torch.uint8 3440377856 0.9075218767253683
{'loss': 1.1536, 'learning_rate': 0.0002, 'epoch': 0.0}
{'loss': 1.2208, 'learning_rate': 0.0002, 'epoch': 0.0}
{'loss': 1.3361, 'learning_rate': 0.0002, 'epoch': 0.0}
{'loss': 1.4817, 'learning_rate': 0.0002, 'epoch': 0.0}
{'loss': 1.8142, 'learning_rate': 0.0002, 'epoch': 0.0}
  0%|                                                                             | 59/38820 [04:16<48:53:15,  4.54s/it]
```

#### Llama2-7b-chat-hf, q4_1, fp32
```
torch.float32 350580736 0.08779149379189381
torch.uint8 3642753024 0.9122085062081062
{'loss': 87.341, 'learning_rate': 0.0002, 'epoch': 0.0}
{'loss': 0.0, 'learning_rate': 0.0002, 'epoch': 0.0}
{'loss': 0.0, 'learning_rate': 0.0002, 'epoch': 0.0}
```

#### Llama2-7b-chat-hf, q4_1, fp32, CPU
```
torch.float32 350580736 0.08779149379189381
torch.uint8 3642753024 0.9122085062081062
{'loss': 87.3288, 'learning_rate': 0.0002, 'epoch': 0.0}
{'loss': 0.0, 'learning_rate': 0.0002, 'epoch': 0.0}
  0%|                                                                        | 20/38820 [1:12:14<2078:46:19, 192.88s/it]
```

#### Llama-2-7B-GPTQ, q4_1, fp32
##### Finetune
```
torch.float32 350588928 0.08779336510974556
torch.uint8 3642753024 0.9122066348902544
{'loss': 78.0657, 'learning_rate': 0.0002, 'epoch': 0.0}
{'loss': 0.0, 'learning_rate': 0.0002, 'epoch': 0.0}
{'loss': 0.0, 'learning_rate': 0.0002, 'epoch': 0.0}
{'loss': 0.0, 'learning_rate': 0.0002, 'epoch': 0.0}
{'loss': 0.0, 'learning_rate': 0.0002, 'epoch': 0.0}
{'loss': 0.0, 'learning_rate': 0.0002, 'epoch': 0.0}
{'loss': 0.0, 'learning_rate': 0.0002, 'epoch': 0.01}
{'loss': 0.0, 'learning_rate': 0.0002, 'epoch': 0.01}
  0%|▏                                                                            | 89/38820 [06:19<38:26:37,  3.57s/it]
```

##### Evaluate (normal):
image

File: python/llm/example/GPU/QLoRA-FineTuning/export_merged_model.py [hunk 1]
```diff
@@ -45,12 +47,37 @@
...
+            print(f"Error: {e}")
+        mid_lora_path = os.path.join(tmpdirname, "adapter_model.bin")
+
+        adapter_path = os.path.join(adapter_path, "adapter_model.bin")
+
+        lora = torch.load(adapter_path, map_location='cpu')
+        tmp_keys = [key[17:-14] for key in lora.keys() if 'lora_A' in key]
+
+        for tmp_key in tmp_keys:
+            a = lora['base_model.model.'+tmp_key+'.lora_A.weight'] / 64
```
Comment (yangw1234) [inline python/llm/example/GPU/QLoRA-FineTuning/export_merged_model.py:75]: Is it possible to use `get_qk_size` instead of 64?

File: python/llm/example/GPU/QLoRA-FineTuning/export_merged_model.py [hunk 2]
```diff
@@ -79,3 +106,6 @@
 
     base_model.save_pretrained(args.output_path, state_dict=deloreanized_sd)
     tokenizer.save_pretrained(args.output_path)
+
+    if qalora and temp_dir:
```
Comment (yangw1234) [inline python/llm/example/GPU/QLoRA-FineTuning/export_merged_model.py:110]: Using `try ... finally ..` ?

File: python/llm/src/bigdl/llm/transformers/qlora.py [hunk 3]
```diff
@@ -140,9 +151,18 @@ def _create_new_module(create_new_module_func, lora_config, adapter_name, target
...
 
 from peft.tuners.lora import LoraModel
+from peft.tuners.lora import LoraConfig as LoraConfigBase
+from dataclasses import dataclass, field
+
+
+@dataclass
+class LoraConfig(LoraConfigBase):
+
+    qa_lora: bool = field(default=False, metadata={"help": "enable qa lora"})
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/qlora.py:161]: qa-lora

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
-# Alpaca QLoRA Finetuning (experimental support)
+# Alpaca QLoRA & QALora Finetuning (experimental support)
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md:1]: QA-LoRA

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md [hunk 5]
```diff
@@ -28,54 +28,75 @@ source /opt/intel/oneapi/setvars.sh
...
 ```
 
-#### Finetuning LLaMA2-7B on four Intel Data Center GPU Max 1550
+##### Finetuning LLaMA2-7B on four Intel Data Center GPU Max 1550
 
 ```bash
 bash finetune_llama2_7b_pvc_1550_4_card.sh
 ```
 
+#### QALora
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md:81]: QA-LoRA

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md [hunk 6]
```diff
@@ -97,3 +118,18 @@ python ./alpaca_qlora_finetuning.py \
...
+
+### 4. Merge the adapter into the original model
+
+#### QLora
+
+```
+python ./export_merged_model.py --repo-id-or-model-path REPO_ID_OR_MODEL_PATH --adapter_path ./outputs/checkpoint-200 --output_path ./outputs/checkpoint-200-merged
+```
+
+#### QALora
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md:130]: QA-LoRA

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md [hunk 7]
```diff
@@ -97,3 +118,18 @@ python ./alpaca_qlora_finetuning.py \
 {'loss': 1.8552, 'learning_rate': 2.9996503623845395e-05, 'epoch': 0.02}                                                                                                                           
   1%|█                                                                                                                                                         | 8/1164 [xx:xx<xx:xx:xx, xx s/it]
 ```
+
+### 4. Merge the adapter into the original model
+
+#### QLora
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md:124]: QLoRA

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md [hunk 8]
```diff
@@ -28,54 +28,75 @@ source /opt/intel/oneapi/setvars.sh
 
 Here, we provide example usages on different hardware. Please refer to the appropriate script based on your device:
 
-#### Finetuning LLaMA2-7B on single Arc A770
+#### QLora
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md:31]: QLoRA

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md [hunk 9]
```diff
@@ -1,6 +1,6 @@
-# Alpaca QLoRA Finetuning (experimental support)
+# Alpaca QLoRA & QALora Finetuning (experimental support)
 
-This example ports [Alpaca-LoRA](https://github.com/tloen/alpaca-lora/tree/main) to BigDL-LLM QLoRA on [Intel GPUs](../../README.md).
+This example ports [Alpaca-LoRA](https://github.com/tloen/alpaca-lora/tree/main) and [QALora](https://github.com/yuhuixu1993/qa-lora/) to BigDL-LLM QLoRA on [Intel GPUs](../../README.md).
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md:3]: This example ports [Alpaca-LoRA]() to BigDL-LLM (using either [QLoRA]() or [QA-LoRA]() algorithm) on [Intel GPU].

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/alpaca_qlora_finetuning.py [hunk 10]
```diff
@@ -109,6 +109,7 @@ def train(
     prompt_template_name: str = "alpaca",  # The prompt template to use, will default to alpaca.
     gradient_checkpointing: bool = False,
     deepspeed: str = None,
+    qalora: bool = False, # if True, use qa-lora https://github.com/yuhuixu1993/qa-lora/
```
Comment (jason-dai) [inline python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/alpaca_qlora_finetuning.py:112]: `qa_lora: bool = False, # if True, use qa-lora https://arxiv.org/abs/2309.04207`

File: python/llm/example/GPU/QLoRA-FineTuning/export_merged_model.py (hunk 1 above)
Comment (jason-dai) [reply python/llm/example/GPU/QLoRA-FineTuning/export_merged_model.py:75]: use a better name such as `get_block_szie`

File: python/llm/example/GPU/QLoRA-FineTuning/export_merged_model.py [hunk 11]
```diff
@@ -45,12 +47,37 @@
                              ', or the path to the huggingface checkpoint folder')
     parser.add_argument('--adapter_path', type=str,)
     parser.add_argument('--output_path', type=str,)
+    parser.add_argument("--qalora", action="store_true", default=False,
```
[the remark this answers is not in the document]
Comment (cyita) [reply python/llm/example/GPU/QLoRA-FineTuning/export_merged_model.py]: Good idea! Updated.

File: python/llm/example/GPU/QLoRA-FineTuning/export_merged_model.py [hunk 12]
```diff
@@ -45,12 +47,37 @@
...
+        try:
+            shutil.copytree(adapter_path, tmpdirname)
+        except OSError as e:
+            print(f"Error: {e}")
+        mid_lora_path = os.path.join(tmpdirname, "adapter_model.bin")
+
+        adapter_path = os.path.join(adapter_path, "adapter_model.bin")
+
+        lora = torch.load(adapter_path, map_location='cpu')
+        tmp_keys = [key[17:-14] for key in lora.keys() if 'lora_A' in key]
```
[the remark this answers is not in the document]
Comment (cyita) [reply python/llm/example/GPU/QLoRA-FineTuning/export_merged_model.py]: Since we only need to change the lora_a weights, I removed key[17:-14] here.

File: python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md (hunk 6 above)
Comment (cyita) [reply python/llm/example/GPU/QLoRA-FineTuning/alpaca-qlora/README.md]: Removed this section and use lora_config.json to determine the adapter type.

**Review by jason-dai — APPROVED** [review]
LGTM

**Comment by cyita** [comment]
The test failed due to FTP maintenance.
